src/main/scala/org/constellation/domain/healthcheck/ReconciliationRound.scala

```diff
@@ -0,0 +1,244 @@
+      isLastRun <- F.delay(attemptsLeft == 0)
+      _ <- if (isFirstRun) F.unit else T.sleep(retryAfter)
+      result <- runReconciliation()
+      _ <- result match {
+        case Left(unresponsivePeers) =>
+          reconciliationResult.modify(
+            // shouldn't we only run healthcheck for nodes that didn't respond instead of running consensus right away?
+            // in consequence I think we should also try to calculate cluster state from the responses we got and start consensuses for these
+            _ => (ReconciliationResult(unresponsivePeers, Set.empty, lastClusterAlignmentResult).some, ())
+          )
+
+        case Right(clusterAlignmentResults) =>
+          if (areAllNodesAligned(clusterAlignmentResults))
+            reconciliationResult.modify(
+              _ => (ReconciliationResult(Set.empty, Set.empty, clusterAlignmentResults.some).some, ())
+            )
+          else if (!isLastRun)
+            run(
+              attemptsLeft - 1,
+              if (isFirstRun) clusterAlignmentResults.some else firstClusterAlignmentResult,
+              clusterAlignmentResults.some
+            )
+          else {
+            for {
+              initiallyMisaligned <- F.delay(
+                firstClusterAlignmentResult.map(fetchMisalignedPeers).getOrElse(Set.empty)
+              )
+              currentlyMisaligned <- F.delay(fetchMisalignedPeers(clusterAlignmentResults))
+              misaligned = initiallyMisaligned.intersect(currentlyMisaligned)
+              _ <- reconciliationResult.modify(
+                _ => (ReconciliationResult(Set.empty, misaligned, clusterAlignmentResults.some).some, ())
+              )
+            } yield ()
+          }
+      }
+    } yield ()
+
+  def areAllNodesAligned(clusterAlignmentResults: Map[Id, List[ClusterAlignmentResult]]): Boolean =
+    clusterAlignmentResults.forall { case (_, result) => result == List(NodeAligned) }
+
+  def fetchMisalignedPeers(clusterAlignmentResults: Map[Id, List[ClusterAlignmentResult]]): Set[Id] =
+    clusterAlignmentResults.filter { case (id, result) => result != List(NodeAligned) }.keySet
+
+  // Should we filter out (not send) node's observation about itself?
+  private def runReconciliation(): F[Either[Set[Id], Map[Id, List[ClusterAlignmentResult]]]] =
+    for {
+      allPeers <- cluster.getPeerInfo
+      peers = allPeers.filter { case (_, peerData) => canActAsJoiningSource(peerData.peerMetadata.nodeState) }
+      // assumption and check before starting reconciliation is that there should be no rounds in progress
+      // if the round starts and reconciliation is in progress, reconciliation should be canceled and result discarded
+      ownClusterState = allPeers.mapValues(
+        pd =>
+          NodeReconciliationData(
+            pd.peerMetadata.id,
+            pd.peerMetadata.nodeState,
+            none[Set[RoundId]],
+            pd.majorityHeight.head.joined // is that correct, head???
```

Comment:
yes, head is the most current and new height is being prepended to that NEL
I don't know though if you need an info about historical heights

Reply:
I don't think I need he historical ones. So I guess it should be fine. Even if we will get historical value combined with e.g. offline status I think it should be ok.